### PR TITLE
Check titandb directory when titan is disabled

### PR DESCRIPTION
**Pull request by tabokie**
Signed-off-by: tabokie <xy.tao@outlook.com>

###  What have you changed?

Check titandb directory is empty (`ls *.blob` count = 0) when titan is disabled. Fatal panic if this check fails, which indicates titan is turned off before all blob files are merged.

###  What is the type of the changes?

- Misc (other changes)

###  How is the PR tested?

- Unit test

### CODE REVIEW

File: cmd/src/server.rs [hunk 1]
```diff
@@ -473,6 +473,21 @@ fn check_system_config(config: &TiKvConfig) {
...
     }
+    // Check blob file dir is empty when titan is disabled
+    if !config.rocksdb.titan.enabled {
+        let titan_dir = if config.rocksdb.titan.dirname.is_empty() {
+            "titandb"
+        } else {
+            config.rocksdb.titan.dirname.as_str()
+        };
+        if let Err(e) = tikv_util::config::check_data_dir_empty(titan_dir) {
+            warn!(
```
Comment (Connor1996) [inline cmd/src/server.rs]: maybe we should refuse to start?

File: components/tikv_util/src/config.rs [hunk 2]
```diff
@@ -754,6 +754,103 @@ pub fn check_data_dir(_data_path: &str) -> Result<(), ConfigError> {
...
+                op, data_path, e
+            ))
+        })?;
+        let mut file_count = 0;
+        for entry in dir {
+            if entry.is_err() {
+                continue;
+            }
+            let entry = entry.unwrap();
+            if entry.path().is_file() {
```
Comment (Connor1996) [inline components/tikv_util/src/config.rs]: count `*.blob` only?

File: cmd/src/server.rs (hunk 1 above)
Comment (yiwu-arbug) [reply cmd/src/server.rs]: +1

File: cmd/src/server.rs [hunk 3]
```diff
@@ -473,6 +473,17 @@ fn check_system_config(config: &TiKvConfig) {
...
     }
+    // Check blob file dir is empty when titan is disabled
+    if !config.rocksdb.titan.enabled {
+        let titan_dir = if config.rocksdb.titan.dirname.is_empty() {
+            "titandb"
+        } else {
+            config.rocksdb.titan.dirname.as_str()
+        };
+        if let Err(e) = tikv_util::config::check_data_dir_empty(titan_dir, "blob") {
+            fatal!("check: titandb-data-dir-empty; err: {}", e);
```
Comment (yiwu-arbug) [inline cmd/src/server.rs]: Shall we '`set_panic_mark` to avoid tikv to restart in this case? @Connor1996

**Comment by tabokie** [comment]
PTAL again @Connor1996 @yiwu-arbug

File: cmd/src/server.rs (hunk 3 above)
Comment (Connor1996) [reply cmd/src/server.rs]: No need, or it's hard for users to run it normally later

Comment (Connor1996) [reply cmd/src/server.rs]: It's better to give a more constructive suggestion, like "titandb data dir is not empty, having enabled Titan before? if you want to disable Titan, please do ...."(you know what I mean)"

File: cmd/src/server.rs [hunk 4]
```diff
@@ -473,6 +474,18 @@ fn check_system_config(config: &TiKvConfig) {
...
     }
+    // Check blob file dir is empty when titan is disabled
+    if !config.rocksdb.titan.enabled {
+        let titan_dir = if config.rocksdb.titan.dirname.is_empty() {
+            "titandb"
+        } else {
+            config.rocksdb.titan.dirname.as_str()
+        };
+        if let Err(e) = tikv_util::config::check_data_dir_empty(titan_dir, "blob") {
+            set_panic_mark();
```
Comment (Connor1996) [inline cmd/src/server.rs]: Usually, `panic_mark` indicates something wrong with the data. Here is just a wrong config.

Comment (tabokie) [reply cmd/src/server.rs]: That sure is one way to interpret it. I will remove it since no **irreversible** corruption has occurred.

**Comment by tabokie** [comment]
> Maybe you can add a test: open Titan and write some data, and disable Titan and check should fail. Then set to fallback and make compact range, then disable Titan and check should pass.

Isn't that supposed to be put in integration test? I don't see how to write a ut for this.

**Comment by Connor1996** [comment]
@tabokie Yes, an integration test. You can write it under `tests/integrations/xxx`.

File: tests/integrations/storage/test_turnoff_titan.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    cluster.must_flush_cf(CF_DEFAULT, true);
+    cluster.compact_data();
+    for path in &titan_paths {
+        assert!(tikv_util::config::check_data_dir_non_empty(path.as_str(), "blob").is_ok());
+    }
+    for i in cluster.get_node_ids().into_iter() {
+        let db = cluster.get_engine(i);
+        let handle = get_cf_handle(&db, CF_DEFAULT).unwrap();
+        let mut opt = Vec::new();
+        opt.push(("blob_run_mode", "kFallback"));
```
Comment (Connor1996) [inline tests/integrations/storage/test_turnoff_titan.rs]: we should disable titan first to make sure the check can fail. and then set to the fallback mode

File: tests/integrations/storage/test_turnoff_titan.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        assert!(db.set_options_cf(handle, &opt).is_ok());
+    }
+    cluster.must_flush_cf(CF_DEFAULT, true);
+    cluster.compact_data();
+    sleep_ms(500);
+    cluster.shutdown();
+    configure_for_disable_titan(&mut cluster);
+    cluster.start_new_engines().unwrap();
```
Comment (Connor1996) [inline tests/integrations/storage/test_turnoff_titan.rs]: why start new engines?

File: tests/integrations/storage/test_turnoff_titan.rs (hunk 5 above)
Comment (tabokie) [reply tests/integrations/storage/test_turnoff_titan.rs]: There is no check here. `check_data_dir_non_empty` is the new check, which doesn't depend on closing titan.

File: tests/integrations/storage/test_turnoff_titan.rs (hunk 6 above)
Comment (tabokie) [reply tests/integrations/storage/test_turnoff_titan.rs]: I intended to check data integrity here.

File: components/tikv_util/src/config.rs [hunk 7]
```diff
@@ -754,6 +754,164 @@ pub fn check_data_dir(_data_path: &str) -> Result<(), ConfigError> {
...
+                    "{}: the number of file with extension {} in directory {} is too much, \
+                     got {}, expect 0.",
+                    op, extension, data_path, count,
+                )));
+            }
+        }
+        Ok(())
+    }
+
+    // check dir is empty of file with certain extension
```
Comment (yiwu-arbug) [inline components/tikv_util/src/config.rs]:
```suggestion
    // check dir is non-empty of file with certain extension
```

File: components/tikv_util/src/config.rs [hunk 8]
```diff
@@ -754,6 +754,164 @@ pub fn check_data_dir(_data_path: &str) -> Result<(), ConfigError> {
...
+    pub fn check_data_dir_empty(data_path: &str, extension: &str) -> Result<(), ConfigError> {
+        let op = "data-dir.empty.check";
+        let dir = Path::new(data_path);
+        if dir.exists() && !dir.is_file() {
+            let count = get_file_count(data_path, extension)?;
+            if count > 0 {
+                return Err(ConfigError::Limit(format!(
+                    "{}: the number of file with extension {} in directory {} is too much, \
```
Comment (yiwu-arbug) [inline components/tikv_util/src/config.rs]:
```suggestion
                    "{}: the number of file with extension {} in directory {} is non-zero, \
```

File: components/tikv_util/src/config.rs [hunk 9]
```diff
@@ -754,6 +754,164 @@ pub fn check_data_dir(_data_path: &str) -> Result<(), ConfigError> {
...
+        let dir = fs::read_dir(data_path).map_err(|e| {
+            ConfigError::FileSystem(format!(
+                "{}: read file dir {:?} failed: {:?}",
+                op, data_path, e
+            ))
+        })?;
+        let mut file_count = 0;
+        for entry in dir {
+            if entry.is_err() {
+                continue;
```
Comment (yiwu-arbug) [inline components/tikv_util/src/config.rs]: should return error in this case?

File: components/test_raftstore/src/util.rs [hunk 10]
```diff
@@ -542,6 +539,46 @@ pub fn create_test_engine(
     (engines, path)
 }
 
+pub fn create_test_engine_on_dir(
```
Comment (Connor1996) [inline components/test_raftstore/src/util.rs]: seems no need anymore

File: components/test_raftstore/src/cluster.rs [hunk 11]
```diff
@@ -127,6 +127,10 @@ impl<T: Simulator> Cluster<T> {
         self.cfg.server.cluster_id
     }
 
+    pub fn pre_start(&mut self) -> result::Result<(), Box<dyn StdError>> {
+        self.cfg.validate()
```
Comment (Connor1996) [inline components/test_raftstore/src/cluster.rs]: can we call it directly in `start`?

Comment (tabokie) [reply components/test_raftstore/src/cluster.rs]: The idea is to check config before `create_engines`

File: components/tikv_util/src/config.rs [hunk 12]
```diff
@@ -754,6 +754,120 @@ pub fn check_data_dir(_data_path: &str) -> Result<(), ConfigError> {
     Ok(())
 }
 
+mod check_data_dir_empty {
```
Comment (Connor1996) [inline components/tikv_util/src/config.rs]: no need to add a mod

File: tests/integrations/storage/test_turnoff_titan.rs [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        let db = cluster.get_engine(i);
+        let handle = get_cf_handle(&db, CF_DEFAULT).unwrap();
+        let mut opt = Vec::new();
+        opt.push(("blob_run_mode", "kFallback"));
+        assert!(db.set_options_cf(handle, &opt).is_ok());
+    }
+    cluster.must_flush_cf(CF_DEFAULT, true);
+    cluster.compact_data();
+    cluster.shutdown();
+    sleep_ms(10000);
```
Comment (Connor1996) [inline tests/integrations/storage/test_turnoff_titan.rs]: sleep too long time, you can put some data to trigger Titan GC before shutdown

**Comment by yiwu-arbug** [comment]
test_turnoff_titan is failing in CI.

File: tests/integrations/storage/test_turnoff_titan.rs [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+            0
+        );
+        assert_eq!(
+            db.get_property_int(&"rocksdb.titandb.num-obsolete-blob-file")
+                .unwrap(),
+            2
+        );
+    }
+    // wait till files are purged.
+    sleep_ms(2000);
```
Comment (Connor1996) [inline tests/integrations/storage/test_turnoff_titan.rs]: too long time

**Comment by tabokie** [comment]
/bench  tidb=pr/12884
```
tidb.yml:
  tikv_client:
    # Max gRPC connections that will be established with each tikv-server.
    grpc-connection-count: 4
```

**Comment by tabokie** [comment]
@yiwu-arbug @Connor1996 PTAL

**Pull request merged**